File: assets/codebase/rendering/AxisTestCameraForwardSpinComponent.cs
namespace gameplay.rendering {
    /// <summary>
    /// Rotates the parent entity around one supplied camera-forward axis using deterministic absolute runtime time.
    /// </summary>
    public sealed class AxisTestCameraForwardSpinComponent : UpdateComponent {
        /// <summary>
        /// Gets or sets the base angle offset in radians applied before time-based rotation.
        /// </summary>
        public float BaseAngleRadians { get; set; }

        /// <summary>
        /// Gets or sets the angular speed in radians per second.
        /// </summary>
        public float AngularSpeedRadians { get; set; }

        /// <summary>
        /// Gets or sets the X component of the camera-forward rotation axis.
        /// </summary>
        public float CameraForwardAxisX { get; set; }

        /// <summary>
        /// Gets or sets the Y component of the camera-forward rotation axis.
        /// </summary>
        public float CameraForwardAxisY { get; set; }

        /// <summary>
        /// Gets or sets the Z component of the camera-forward rotation axis.
        /// </summary>
        public float CameraForwardAxisZ { get; set; }

        /// <summary>
        /// Evaluates the current orientation from total elapsed runtime time.
        /// </summary>
        public override void Update() {
            base.Update();

            double angleRadians = BaseAngleRadians + (AngularSpeedRadians * Core.Instance.TotalElapsedSeconds);
            float3 axis = float3.Normalize(new float3(CameraForwardAxisX, CameraForwardAxisY, CameraForwardAxisZ));

            float4 orientation;
            float4.CreateFromAxisAngle(ref axis, (float)angleRadians, out orientation);
            orientation.Normalize();
            Parent.LocalOrientation = orientation;
        }
    }
}

File: windows-build-20260602-stack-box-direct-real-bepu-fix1/cooked/code/gameplay/AxisTestCameraForwardSpinComponent.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "AxisTestCameraForwardSpinComponent.hpp"
#include "AxisTestCameraForwardSpinComponent.hpp"
#include "Component.hpp"
#include "Core.hpp"
#include "Entity.hpp"
#include "UpdateComponent.hpp"
#include "float3.hpp"
#include "float4.hpp"

AxisTestCameraForwardSpinComponent::AxisTestCameraForwardSpinComponent() : AngularSpeedRadians(), BaseAngleRadians(), CameraForwardAxisX(), CameraForwardAxisY(), CameraForwardAxisZ()
{
}

float AxisTestCameraForwardSpinComponent::get_AngularSpeedRadians()
{
return this->AngularSpeedRadians;
}

void AxisTestCameraForwardSpinComponent::set_AngularSpeedRadians(float value)
{
this->AngularSpeedRadians = value;
}

float AxisTestCameraForwardSpinComponent::get_BaseAngleRadians()
{
return this->BaseAngleRadians;
}

void AxisTestCameraForwardSpinComponent::set_BaseAngleRadians(float value)
{
this->BaseAngleRadians = value;
}

float AxisTestCameraForwardSpinComponent::get_CameraForwardAxisX()
{
return this->CameraForwardAxisX;
}

void AxisTestCameraForwardSpinComponent::set_CameraForwardAxisX(float value)
{
this->CameraForwardAxisX = value;
}

float AxisTestCameraForwardSpinComponent::get_CameraForwardAxisY()
{
return this->CameraForwardAxisY;
}

void AxisTestCameraForwardSpinComponent::set_CameraForwardAxisY(float value)
{
this->CameraForwardAxisY = value;
}

float AxisTestCameraForwardSpinComponent::get_CameraForwardAxisZ()
{
return this->CameraForwardAxisZ;
}

void AxisTestCameraForwardSpinComponent::set_CameraForwardAxisZ(float value)
{
this->CameraForwardAxisZ = value;
}

void AxisTestCameraForwardSpinComponent::Update()
{
UpdateComponent::Update();
const double angleRadians = this->BaseAngleRadians + (this->AngularSpeedRadians * Core::get_Instance()->get_TotalElapsedSeconds());
float3 axis = float3::Normalize(float3(this->CameraForwardAxisX, this->CameraForwardAxisY, this->CameraForwardAxisZ));
float4 orientation;
float4::CreateFromAxisAngle__ref0_out2(axis, static_cast<float>(angleRadians), orientation);
orientation.Normalize();
this->Parent->set_LocalOrientation(orientation);
}

